### PR TITLE
We use InternLM Auto Formalizer formalize 20 IMO problems.

### PR DESCRIPTION
**Pull request by objecti0n**
The description of InternLM Auto Formalizer is in [link](https://github.com/InternLM/InternLM-Math/tree/main/leanworkbook).
Three human experts check and modify the autoformalization.
We notice our autoformalization does not have **solution_set**, maybe we can merge pure-proof problems.

These 20 problems have all been compiled using following code:
```
$ lake exe cache get
$ lake build
```

### CODE REVIEW

File: Compfiles/Imo1959P4.lean [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.Algebra] }
+
+/-!
+# International Mathematical Olympiad 1959, Problem 4
+
+Construct a right triangle with a given hypotenuse $c$ such that the median drawn to the hypotenuse is the geometric mean of the two legs of the triangle.
```
Comment (dwrensha) [inline Compfiles/Imo1959P4.lean:16]: My understanding is that "construct" in this context means [straightedge and compass construction](https://en.wikipedia.org/wiki/Straightedge_and_compass_construction). So framing this as a purely algebraic problem misses a big part of what's being asked for.

File: Compfiles/Imo1978P1.lean [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.NumberTheory] }
+
+/-!
+# International Mathematical Olympiad 1978, Problem 1
+
+m and n are positive integers with m < n. The last three decimal digits of 1978m are the same as the last three decimal digits of 1978n. Find m and n such that m + n has the least possible value.
```
Comment (dwrensha) [inline Compfiles/Imo1978P1.lean:16]: "Find m and n" means that this problem should have a `determine` statement.

File: Compfiles/Imo1979P5.lean [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.Algebra] }
+
+/-!
+# International Mathematical Olympiad 1979, Problem 5
+
+Find all real numbers a for which there exist non-negative real numbers x1, x2, x3, x4, x5 satisfying:
```
Comment (dwrensha) [inline Compfiles/Imo1979P5.lean:16]: "Find all" means that this problem should have a `determine` statement.

File: Compfiles/Imo1988P3.lean [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import ProblemExtraction
+
+problem_file { tags := [.Algebra] }
+
+/-!
+# International Mathematical Olympiad 1988, Problem 3
+
+A function $f$ defined on the positive integers (and taking positive integers values) is given by:
+$\begin{matrix} f(1) = 1, f(3) = 3 \\ f(2 \cdot n) = f(n) \\ f(4 \cdot n + 1) = 2 \cdot f(2 \cdot n + 1) - f(n) \\ f(4 \cdot n + 3) = 3 \cdot f(2 \cdot n + 1) - 2 \cdot f(n), \end{matrix}$
+for all positive integers $n.$ Determine with proof the number of positive integers $\leq 1988$ for which $f(n) = n.$
```
Comment (dwrensha) [inline Compfiles/Imo1988P3.lean:18]: This should have a `determine` statement.

File: Compfiles/Imo1994P5.lean [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.Algebra] }
+
+/-!
+# International Mathematical Olympiad 1994, Problem 5
+
+Let S be the set of all real numbers greater than -1. Find all functions f :S→S such that f(x + f(y) + xf(y)) = y + f(x) + yf(x) for all x and y, and f(x)/x is strictly increasing on each of the intervals -1 < x < 0 and 0 < x.
```
Comment (dwrensha) [inline Compfiles/Imo1994P5.lean:16]: "Find all" means this should have a `determine` statement.

File: Compfiles/Imo1995P4.lean [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+problem_file { tags := [.Algebra] }
+
+/-!
+# International Mathematical Olympiad 1995, Problem 4
+
+The positive real numbers $x_0, x_1, x_2,.....x_{1994}, x_{1995}$ satisfy the relations
+$x_0=x_{1995}$ and $x_{i-1}+\frac{2}{x_{i-1}}=2{x_i}+\frac{1}{x_i}$
+for $i=1,2,3,....1995$
+Find the maximum value that $x_0$ can have.
```
Comment (dwrensha) [inline Compfiles/Imo1995P4.lean:19]: "Find" means this should have a `deteremine` statement.

File: Compfiles/Imo1998P3.lean [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.NumberTheory] }
+
+/-!
+# International Mathematical Olympiad 1998, Problem 3
+
+For any positive integer $n$, let $d(n)$ denote the number of positive divisors of $n$ (including 1 and $n$ itself). Determine all positive integers $k$ such that $d(n^2)/d(n) = k$ for some $n$.
```
Comment (dwrensha) [inline Compfiles/Imo1998P3.lean:16]: Should have a `determine` statement.

File: Compfiles/Imo1998P4.lean [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.NumberTheory] }
+
+/-!
+# International Mathematical Olympiad 1998, Problem 4
+
+Determine all pairs (a, b) of positive integers such that ab^2 + b + 7 divides a^2b + a + b.
```
Comment (dwrensha) [inline Compfiles/Imo1998P4.lean:16]: Should have a `deteremine` statement.

File: Compfiles/Imo1998P6.lean [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.Algebra] }
+
+/-!
+# International Mathematical Olympiad 1998, Problem 6
+
+Consider all functions f from the set of all positive integers into itself satisfying f(t^2f(s)) = sf(t)^2 for all s and t. Determine the least possible value of f(1998).
```
Comment (dwrensha) [inline Compfiles/Imo1998P6.lean:16]: Should have a `determine` statement.

File: Compfiles/Imo1959P4.lean [hunk 10]
```diff
@@ -0,0 +1,21 @@
+/-
+Copyright (c) 2024 The Compfiles Contributors. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: InternLM-MATH LEAN Formalizer v0.1. This is auto-formalized by InternLM-MATH LEAN Formalizer v0.1, modified and verified by InternLM-MATH team members.
```
Comment (dwrensha) [inline Compfiles/Imo1959P4.lean:4]: The `Authors` line should be a comma-separated list of authors and not include extra explanations. We parse this list for display on the compfiles web page: https://github.com/dwrensha/compfiles/blob/f99bf6f2928d47dd1a445b414b3a723c2665f091/ProblemExtraction.lean#L133-L139

Just "InternLM-MATH LEAN Formalizer v0.1" would be OK.

File: Compfiles/Imo1963P4.lean [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+/-!
+# International Mathematical Olympiad 1963, Problem 4
+
+Find all solutions $x_1,x_2,x_3,x_4,x_5$ of the system \begin{eqnarray*} x_5+x_2&=&yx_1\\ x_1+x_3&=&yx_2\\ x_2+x_4&=&yx_3\\ x_3+x_5&=&yx_4\\ x_4+x_1&=&yx_5,\end{eqnarray*} where $y$ is a parameter.
+-/
+
+namespace Imo1963P4
+
+determine solution_set : Set (ℝ × ℝ × ℝ × ℝ × ℝ) := {(a, b, c, d, e) | a + b + c + d + e = 0 ∨ (a = b ∧ b = c ∧ c = d ∧ d = e)}
```
Comment (dwrensha) [inline Compfiles/Imo1963P4.lean:21]: This `solution_set` needs to depend on `y`.
```lean
determine solution_set (y : ℝ) : Set (ℝ × ℝ × ℝ × ℝ × ℝ) := ...
```

File: Compfiles/Imo1972P4.lean [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+(x_3^2 - x_5x_2)(x_4^2 - x_5x_2) ≤ 0
+(x_4^2 - x_1x_3)(x_5^2 - x_1x_3) ≤ 0
+(x_5^2 - x_2x_4)(x_1^2 - x_2x_4) ≤ 0
+-/
+
+namespace Imo1972P4
+
+determine solution_set : Set (ℝ × ℝ × ℝ × ℝ × ℝ) := {(a, b, c, d, e) | a = b ∧ b = c ∧ c = d ∧ d = e}
+
+problem imo1972_p4 (a b c d e : ℝ) (h₀ : 0 < a ∧ 0 < b ∧ 0 < c ∧ 0 < d ∧ 0 < e) (h₁ : (a^2 - c * e) * (b^2 - c * e) ≤ 0) (h₂ : (b^2 - d * a) * (c^2 - d * a) ≤ 0) (h₃ : (c^2 - e * b) * (d^2 - e * b) ≤ 0) (h₄ : (d^2 - a * c) * (e^2 - a * c) ≤ 0) (h₅ : (e^2 - b * d) * (a^2 - b * d) ≤ 0) : (a, b, c, d, e) ∈ solution_set := by sorry
```
Comment (dwrensha) [inline Compfiles/Imo1972P4.lean:29]: This only proves one direction. You also need the reverse implication as well: if (a,b,c,d,e) is in the solution set, then the given equations hold.

Comment (dwrensha) [reply Compfiles/Imo1972P4.lean:29]: Usually we set up this kind of problem with an `Iff` (`↔`).

File: Compfiles/Imo1978P1.lean (hunk 2 above)
Comment (dwrensha) [reply Compfiles/Imo1978P1.lean:16]: No, it's not asking for a `Set`. It's asking for just two numbers.
```lean
determine solution : ℕ × ℕ := (3, 103)
```

File: Compfiles/Imo1978P6.lean [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+/-!
+# International Mathematical Olympiad 1978, Problem 6
+
+An international society has its members from six different countries. The list of members has 1978 names, numbered $1, 2, \ldots, 1978$. Prove that there is at least one member whose number is the sum of the numbers of two (not necessarily distinct) members from his own country.
+$
+-/
+
+namespace Imo1978P6
+
+problem imo1978_p6 (n : ℕ) (hn : n = 1978) (C : Fin n → Fin 6) : ∃ i : Fin n, ∃ j : Fin n, ∃ k : Fin n, C i = C j ∧ C j = C k ∧ i ≠ k ∧ (i:ℕ) + (k:ℕ) = (j:ℕ) + 1 := by sorry
```
Comment (dwrensha) [inline Compfiles/Imo1978P6.lean:22]: This does not look right. I guess `j` is the member we're looking for, but then why do we have `i ≠ k`? The problem specifically says that the two other members are "not necessarily distinct".

Also, the problem says the numbers go from 1 to 1978, but the numbers in this formalization go from 0 to 1977. To correct for that, I think the equality needs to be `(i:ℕ) + 1 + (k:ℕ) + 1 = (j:ℕ)  + 1 ` or `(i:ℕ) + (k:ℕ) + 1 = (j:ℕ) `

File: Compfiles/Imo1979P5.lean [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+x1 + 2x_2 + 3x_3 + 4x_4 + 5x_5 = a,
+x1 + 2^3x_2 + 3^3x_3 + 4^3x_4 + 5^3x_5 = a^2,
+x1 + 2^5x_2 + 3^5x_3 + 4^5x_4 + 5^5x_5 = a^3.
+-/
+
+namespace Imo1979P5
+
+determine solution_set : Set ℝ := {0, 1, 4, 9, 16, 25}
+
+problem imo1979_p5 (a : ℝ) : (∃ x1 x2 x3 x4 x5 : ℝ, x1 ≥ 0 ∧ x2 ≥ 0 ∧ x3 ≥ 0 ∧ x4 ≥ 0 ∧ x5 ≥ 0 ∧ x1 + 2*x2 + 3*x3 + 4*x4 + 5*x5 = a ∧ x1 + 2^3*x2 + 3^3*x3 + 4^3*x4 + 5^3*x5 = a^2 ∧ x1 + 2^5*x2 + 3^5*x3 + 4^5*x4 + 5^5*x5 = a^3 ) ↔ a ∈ solution_set := by sorry
```
Comment (dwrensha) [inline Compfiles/Imo1979P5.lean:27]: This line and a lot of others in the PR are very long. Please limit to 100 characters.

File: Compfiles/Imo1971P3.lean [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+/-!
+# International Mathematical Olympiad 1971, Problem 3
+
+Prove that we can find an infinite set of positive integers of the form 2^n - 3 (where n is a positive integer) every pair of which are relatively prime.
+-/
+
+namespace Imo1971P3
+
+problem imo1971_p3 : ∀ M : ℕ, ∃ n m : ℕ, n > m ∧ m > M ∧ Nat.Coprime (2 ^ n - 3) (2 ^ m - 3) := by sorry
```
[the remark this answers is not in the document]
Comment (alreadydone) [reply Compfiles/Imo1971P3.lean]: Notice that the classic infinitude of primes in Lean uses unboundedness in the form of `∀ N, ∃ p >= N, prime p`

File: Compfiles/Imo1963P4.lean [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+x_4+x_1&=&yx_5
+
+where $y$ is a parameter.
+-/
+
+namespace Imo1963P4
+
+determine solution_set : Set (ℝ × ℝ × ℝ × ℝ × ℝ × ℝ) := sorry
+
+problem imo1963_p4 (x : Fin 6 → ℝ) (y : ℝ)
```
Comment (dwrensha) [inline Compfiles/Imo1963P4.lean:31]: This still has the following issues:
1. `solution_set` should depend on `y` (not include `y` in the product, as you have done)
2. It's awkward to use `Fin 6` when there are only 5 variables x_i.
3. Still need both directions for if-and-only-if.
4. Some of the indexes for x in the formalized problem statement are wrong.

I decided that writing my own formalization would be the easiest way forward here: https://github.com/dwrensha/compfiles/commit/8a33ae45e8260032dd7858fccb9c016e47a6a97a

**Comment by dwrensha** [comment]
Thanks!
Rebased and squashed in https://github.com/dwrensha/compfiles/commit/d624eddb3ce2294cb52bf6f165913ea62c9b3155.

File: Compfiles/Imo1998P4.lean [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+-/
+
+namespace Imo1998P4
+
+determine solution_set : Set (ℕ × ℕ) := {(11, 1), (49, 1)} ∪
+  {(x,y) | ∃ k : ℕ , (x = 7 * k^2 ∧ y = 7 * k)}
+
+problem imo1998_p4 (a b : ℕ) :
+  a * b^2 + b + 7 ∣ a^2 * b + a + b ↔ (a, b) ∈ solution_set := by sorry
```
Comment (dwrensha) [inline Compfiles/Imo1998P4.lean:25]: I added the condition that `a` and `b` must be positive: https://github.com/dwrensha/compfiles/commit/83d5f07d33a7c3a685cb16df5ad6170d0f50d747

File: Compfiles/Imo1998P6.lean [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+f(t^2f(s)) = sf(t)^2 for all s and t.
+Determine the least possible value of f(1998).
+-/
+
+namespace Imo1998P6
+
+determine solution : ℕ := 120
+
+problem imo1998_p6
+  (f : ℕ → ℕ)
```
Comment (dwrensha) [inline Compfiles/Imo1998P6.lean:26]: I changed this from `Nat` to `PNat` in https://github.com/dwrensha/compfiles/commit/aabfda06bd19abb64f9815126f18ca749447daa1

**Pull request closed without merging**